File: srtf.py
import matplotlib.pyplot as plt

class Process:
    def __init__(self, pid, arrival, burst):
        self.pid = pid
        self.arrival = arrival
        self.burst = burst
        self.remaining = burst
        self.completion = 0
        self.start = -1

n = int(input("Enter number of processes: "))
processes = []
for i in range(n):
    pid = f'P{i+1}'
    arrival = int(input(f"Enter arrival time for {pid}: "))
    burst = int(input(f"Enter burst time for {pid}: "))
    processes.append(Process(pid, arrival, burst))

time = 0
completed = 0
gantt = []

while completed < n:
    ready = [p for p in processes if p.arrival <= time and p.remaining > 0]
    
    if ready:
        current = min(ready, key=lambda x: x.remaining)
        if current.start == -1:
            current.start = time
        gantt.append(current.pid)
        current.remaining -= 1
        time += 1
        if current.remaining == 0:
            current.completion = time
            completed += 1
    else:
        gantt.append('Idle')
        time += 1

print("\nProcess\tAT\tBT\tCT\tTAT\tWT")
total_tat = total_wt = 0
for p in processes:
    tat = p.completion - p.arrival
    wt = tat - p.burst
    total_tat += tat
    total_wt += wt
    print(f"{p.pid}\t{p.arrival}\t{p.burst}\t{p.completion}\t{tat}\t{wt}")

print(f"\nAverage TAT: {total_tat/n:.2f}")
print(f"Average WT: {total_wt/n:.2f}")

plt.figure(figsize=(10, 2))
start = 0
for pid in gantt:
    plt.barh(0, 1, left=start, edgecolor='black')
    plt.text(start + 0.5, 0, pid, ha='center', va='center')
    start += 1

plt.yticks([])
plt.xlabel("Time")
plt.title("Gantt Chart - SRTF Scheduling")
plt.show()

File: srtf.cpp
#include <iostream>
#include <climits>
#include <iomanip>
#include <vector>

using namespace std;

struct Process
{
    int pid;
    int at;
    int bt;
};

int main()
{
    int n;
    cout << "Enter total number of processes: ";
    cin >> n;

    Process proc[n];
    int rt[n];
    int wt[n], tat[n], ct[n];
    float total_wt = 0, total_tat = 0;
    vector<int> timeline; // For Gantt Chart

    for (int i = 0; i < n; i++)
    {
        proc[i].pid = i + 1;
        cout << "Enter AT and BT for Process " << i + 1 << ": ";
        cin >> proc[i].at >> proc[i].bt;
        rt[i] = proc[i].bt;
    }

    int completed = 0, currentTime = 0;
    while (completed != n)
    {
        int shortest = -1;
        int min_rt = INT_MAX;

        for (int j = 0; j < n; j++)
        {
            if ((proc[j].at <= currentTime) && (rt[j] < min_rt) && rt[j] > 0)
            {
                min_rt = rt[j];
                shortest = j;
            }
        }

        if (shortest == -1)
        {
            timeline.push_back(0); // 0 for IDLE
            currentTime++;
        }
        else
        {
            timeline.push_back(proc[shortest].pid);
            rt[shortest]--;

            if (rt[shortest] == 0)
            {
                completed++;
                ct[shortest] = currentTime + 1;
                tat[shortest] = ct[shortest] - proc[shortest].at;
                wt[shortest] = tat[shortest] - proc[shortest].bt;

                total_wt += wt[shortest];
                total_tat += tat[shortest];
            }
            currentTime++;
        }
    }

    cout << "\nPID\tAT\tBT\tCT\tTAT\tWT" << endl;
    for (int i = 0; i < n; i++)
    {
        cout << proc[i].pid << "\t" << proc[i].at << "\t"
             << proc[i].bt << "\t" << ct[i] << "\t"
             << tat[i] << "\t" << wt[i] << endl;
    }

    cout << "\n--- Gantt Chart ---" << endl;
    int current_pid = timeline[0];
    int start_time = 0;

    for (int t = 1; t < timeline.size(); t++)
    {
        if (timeline[t] != current_pid)
        {
            cout << start_time << "--" << t << "\t| ";
            if (current_pid == 0)
            {
                cout << "IDLE";
            }
            else
            {
                cout << "P" << current_pid;
            }
            cout << " |" << endl;

            current_pid = timeline[t];
            start_time = t;
        }
    }

    cout << start_time << "--" << timeline.size() << "\t| ";
    if (current_pid == 0)
    {
        cout << "IDLE";
    }
    else
    {
        cout << "P" << current_pid;
    }
    cout << " |" << endl;

    cout << "\nAverage Waiting Time: " << (total_wt / n) << endl;
    cout << "Average Turnaround Time: " << (total_tat / n) << endl;

    return 0;
}
